Implements a first draft of a validator Kotlin DSL

hartmut-co-uk opened the pull request:
### State
This is a first small step to be considered for review/discussion purpose.

### Notes
- CharSequenceConstraint support only to get started
- no support for 'ifPresent' vs. 'required'
- no support for 
  - konstraintOnCondition
  - konstraintOnGroup
  - konstraintOnObject
  - nest
  - forEach
  - override messages
  - ...

### Next Steps
- confirm approach, alternative options to defining a separate `class ValidatorBuilderKt<T> : ValidatorBuilder<T>`?
- define complete Kotlin DSL covering more/all YAVI features

### TODOs
- [ ] update https://github.com/making/yavi-doc/blob/main/src/main/asciidoc/using-yavi.adoc

hartmut-co-uk commented:
as mentioned in https://github.com/making/yavi/issues/18 I pretty much followed the DSL of **konform**.
The [Advanced use](https://github.com/konform-kt/konform#advanced-use) section also has more good examples how the DSL could be further specced out.

making commented:
@hartmut-co-uk Thanks! looks cool.

Is it possible to use delegation instead of inheritance?
I want to avoid for `Validator` and its derivations relying on `ValidatorBuilder`.

making commented:
@hartmut-co-uk  I got feedback on [Twitter](https://twitter.com/toliner_/status/1410056757301288964?s=20) from @toliner. I translate it here

1. How about inlining higher-order functions in DSL from a performance perspective?
2. It could be a user's preference, but within the scope of `ValidateBuilderKt`, it is preferable to be able to read and write like a sentence by using the infix function like [kotest](https://github.com/kotest/kotest) as follows.
This can be achieved by adding an infix extension function.
```kotlin
DemoString::x {
    it greatherThan 1
    it lessThan 5
}
```

hartmut-co-uk commented:
**Progress** 
- `ValidatorKt` _class_ has been replaced with a `inline fun <T> validator(...)`
- added first infix fun _required_ (also impl. + working for CharSequence)
- added DSL infix functions _nest_ & _forEach_ - but wasn't able to implement due to the weird setup of `ValidatorBuilderKt<T>` extending Java class `ValidatorBuilder<T>` and the need to put the infix KProperty extension functions inside so they have access to `this` (ValidatorBuilderKt). Help needed. ping @sdeleuze who opened https://github.com/making/yavi/issues/18 2y back :-)

Please see DSL samples below..

**new validator<> fun**
```
val validator = validator<DemoString> {
    DemoString::x {
        notEmpty()
        lessThan(5)
    }
}
```

**required infix**
```
val validator = validator<DemoString> {
    DemoString::x required {
        lessThan(5)
    }
}
```

**forEach**
```
val validator = validator<DemoForEachCollection> {
    DemoForEachCollection::x forEach {
        DemoString::x {
            greaterThan(1)
            lessThan(5)
        }
    }
}
```

**nest**
```
val validator = validator<DemoNested> {
    DemoNested::x nest {
        DemoString::x { greaterThan(1).lessThan(5) }
    }
}
```

hartmut-co-uk commented on src/main/kotlin/am/ik/yavi/builder/ValidatorBuilderExtensions.kt:
```diff
@@ -177,4 +177,34 @@ fun <T> ValidatorBuilder<T>.konstraintOnCondition(condition: ConstraintCondition
+        infix fun KProperty1<T, String?>.required(block: CharSequenceConstraint<T, String?>.() -> Unit) =
+                constraint(this, this.name) { it.notNull().apply(block) }
+
+        infix fun <L: Collection<E>?, E> KProperty1<T, L?>.forEach(block: ValidatorBuilderKt<E>.() -> Unit) {
+                TODO()
+                /** can't get following to work due to cumbersome inheritance setup of ValidatorBuilderKt :-( */
+                // forEach(this, this.name) { it.apply(block) }
+        }
+
+        infix fun <N> KProperty1<T, N?>.nest(block: ValidatorBuilderKt<N>.() -> Unit) {
+                // TODO()
+                /** also not possible to cast... :-(( */
+                // nest(this, this.name) { it.apply(block as ValidatorBuilder<N?>.() -> Unit) }
+        }
```
Would anyone be able to help unblock / suggest a way forward?
Are there any options without wider refactoring of Yavi itself? 
(and without moving/rewriting Yavi in Kotlin 😅)

ngsw-taro replied:
```kotlin
inline fun <T> validator(init: ValidatorBuilderKt<T>.() -> Unit): Validator<T> {
        val builder = ValidatorBuilderKt<T>(ValidatorBuilder())
        return builder.apply(init).build()
}

class ValidatorBuilderKt<T>(private val validatorBuilder: ValidatorBuilder<T>) {

        fun build(): Validator<T> {
            return validatorBuilder.build()
        }

        /** Q: this will be required for each type, right? */
        operator fun KProperty1<T, String?>.invoke(block: CharSequenceConstraint<T, String?>.() -> Unit) =
                validatorBuilder.constraint(this, this.name) { it.apply(block) }

        /** Q: this will be required for each type, right? */
        infix fun KProperty1<T, String?>.required(block: CharSequenceConstraint<T, String?>.() -> Unit) =
                validatorBuilder.constraint(this, this.name) { it.notNull().apply(block) }

        infix fun <L: Collection<E>?, E> KProperty1<T, L?>.forEach(block: ValidatorBuilderKt<E>.() -> Unit) {
                validatorBuilder.forEach(this, this.name) { ValidatorBuilderKt(it).apply(block).validatorBuilder }
        }

        infix fun <N> KProperty1<T, N>.nest(block: ValidatorBuilderKt<N>.() -> Unit) {
               validatorBuilder.nest(this, this.name) { ValidatorBuilderKt(it).apply(block).validatorBuilder }
        }

        infix fun <N> KProperty1<T, N?>.nest(validator: Validator<N>): ValidatorBuilder<T> = TODO()
}
```

hartmut-co-uk replied:
brilliant, thank you! ✨ 
looks like a solid pattern to proceed with...

hartmut-co-uk commented:
@making with @ngsw-taro's provided fix I think we can now proceed.
If you're happy with current code I'd continue to add coverage for all missing features and constraints.

What about the `konstraint` extension functions? 
Would you want to keep them or drop in favour of this PR's DSL?

making commented:
@hartmut-co-uk Let's deprecate `konstraint`

hartmut-co-uk commented:
(ready to be reviewed)

making commented:
@hartmut-co-uk Great work! looks pretty cool!

Is it possible to add an option to explicitly specify the fieldName of constarint?


```kotlin
    @JvmName("invokeTString")
    operator fun KProperty1<T, String?>.invoke(block: CharSequenceConstraint<T, String?>.() -> Unit) =
        validatorBuilder.constraint(this, this.name) { it.apply(block) }
```

Sometimes I may want to use a custom field name instead of `this.name` above.
I'm not sure what the DSL would look like.

hartmut-co-uk commented:
something like this?
```
val validator = validator<DemoString> {
    DemoString::x {
        notEmpty()
        lessThan(5)
    }
}
```
-> named
```
val validator2 = validator<DemoString> {
    (DemoString::x)("test") {
        notEmpty()
        lessThan(5)
    }
}
```

with 
```
    @JvmName("invokeTString")
    operator fun KProperty1<T, String?>.invoke(name: String, block: CharSequenceConstraint<T, String?>.() -> Unit) =
        validatorBuilder.constraint(this, name) { it.apply(block) }
```

making commented:
@hartmut-co-uk Yes, something like that

hartmut-co-uk commented on src/test/kotlin/am/ik/yavi/builder/ValidatorBuilderKtDslTest.kt:
```diff
@@ -0,0 +1,1032 @@
+
+    @Test
+    fun constraintOnCharSequenceRequired() {
+        val validator = validator<DemoString> {
+            DemoString::x required {
+                lessThan(5)
+            }
+        }
```
how about this _required_ infix - since this is something 'additional' to the non-Kotlin-DSL can this also be added? (for all supported constraint types)

Implementation is:
```kotlin
    /** Q: this will be required for each type */
    infix fun KProperty1<T, String?>.required(block: CharSequenceConstraint<T, String?>.() -> Unit) =
        validatorBuilder.constraint(this, this.name) { it.notNull().apply(block) }
```

If I my understanding is right, since YAVI is Java first - by default all validations are _optional_ and will only be applied if the field is notNull.

making replied:
@ngsw-taro any thoughts?

making replied:
> by default all validations are optional and will only be applied if the field is notNull.

Except for `notNull`, `notBlank` and `notEmpty`, `null` is regarded as a valid input 
https://yavi.ik.am/#built-in-constraints

hartmut-co-uk replied:
it's quite experimental and different from YAVI itself, maybe better to just drop the idea?

It's obviously straightforward and clean to just do the normal way...

```kotlin
val validator = validator<DemoString> {
    DemoString::x {
        notNull()
        lessThan(5)
    }
}
```

making replied:
From a non-Kotlin user perspective, I prefer the normal way ;)

hartmut-co-uk replied:
👍  I have removed the infix fun + test

hartmut-co-uk commented:
> @hartmut-co-uk Yes, something like that

I can't think of any better/other way to achieve this - do you want me to add functions for all types?

making commented:
> > @hartmut-co-uk Yes, something like that
> 
> I can't think of any better/other way to achieve this - do you want me to add functions for all types?

I can add if you don't want to add **all types** 😅

hartmut-co-uk commented:
I'm happy to do so, just wanted to get your OK before moving forward.

hartmut-co-uk commented:
> I'm happy to do so, just wanted to get your OK before moving forward.

done

making commented:
It looks like we can merge this PR.
I will merge this after releasing 0.8.3 that fixes gh-156 .

making commented:
@hartmut-co-uk 
Merged, thanks for you great work!
